pip/main.ts

```diff
@@ -0,0 +1,158 @@
+const subcommand = Deno.args[0];
+if (Deno.args.length != 0 && !knownSubcommands.includes(subcommand)) {
+  console.error(`[${red("phylum")}] This extension does not support arguments before the first subcommand. Please open an issue if "${subcommand}" is not an argument.`);
+}
+
+// Ignore all commands that shouldn't be intercepted.
+if (Deno.args.length == 0 || subcommand != "install") {
+  const cmd = Deno.run({ cmd: ["pip", ...Deno.args] });
+  const status = await cmd.status();
+  Deno.exit(status.code);
+}
+
+// Analyze new dependencies with phylum before install/update.
+await checkDryRun();
+
+// Perform the package installation.
+const cmd = Deno.run({ cmd: ["pip", ...Deno.args] });
+const status = await cmd.status();
+Deno.exit(status.code);
+
+// Analyze new packages.
+async function checkDryRun() {
+  console.log(`[${green("phylum")}] Finding new dependencies…`);
+
+  const status = PhylumApi.runSandboxed({
+    cmd: "pip",
+    args: [...Deno.args, "--dry-run"],
+    exceptions: {
+      run: ["/bin", "/usr/bin", "~/.pyenv"],
+      write: ["~/Library/Caches/pip", "~/.cache", "~/.local/lib", "/tmp"],
+      read: ["~/Library/Caches/pip", "~/.cache", "~/.local/lib", "/tmp", "/etc/passwd", "/etc/apache2/mime.types"],
+      net: true,
+    },
+    stdout: 'piped',
+  });
+
+  // Ensure dry-run was successful.
+  if (!status.success) {
+    console.error(`[${red("phylum")}] Pip dry-run failed.\n`);
+    await Deno.exit(status.code);
+  }
+
+  // Parse dry-run output.
+  let packages = parseDryRun(status.stdout);
+
+  console.log(`[${green("phylum")}] Dependency resolution successful.\n`);
+  console.log(`[${green("phylum")}] Analyzing packages…`);
+
+  if (packages.length === 0) {
+    console.log(`[${green("phylum")}] No new packages found for analysis.\n`);
+    return;
+  }
+
+  const jobId = await PhylumApi.analyze("pypi", packages);
+  const jobStatus = await PhylumApi.getJobStatus(jobId);
+
+  if (jobStatus.pass && jobStatus.status === "complete") {
+    console.log(`[${green("phylum")}] All packages pass project thresholds.\n`);
+  } else if (jobStatus.pass) {
+    console.warn(
+      `[${
+        yellow(
+          "phylum",
+        )
+      }] Unknown packages were submitted for analysis, please check again later.\n`,
+    );
+    Deno.exit(126);
+  } else {
+    console.error(
+      `[${red("phylum")}] The operation caused a threshold failure.\n`,
+    );
+    Deno.exit(127);
+  }
+}
+
+// Parse the dry-run output of `pip install`.
+function parseDryRun(output: string): [object] {
+  // Extract the "Would install [..]" line.
+  let new_deps;
+  const lines = output.split('\n');
+  for (var i = 0; i < lines.length; i++) {
+    if (lines[i].startsWith('Would install ')) {
+      new_deps = lines[i].substring('Would install '.length);
+      break;
+    }
+  }
+
+  // No "Would install [..]" means there were no new dependencies.
+  if (!new_deps) {
+    return [];
```

Comment:
Is it possible we could miss the "Would install"? Maybe it's a different version of pip? Should we output a message saying we don't think there are packages to install and exit 0 in this case without continuing to avoid potential mistakes?

Reply:
I am not familiar with historic pip versions, but no "would install" is standard for when no updates need to be installed. In this case we do also print a message.